providers/gpgpu/bin/gpu_passthrough.py

```diff
@@ -0,0 +1,576 @@
+#!/usr/bin/env python3
+"""Script to test virtualization GPU passthrough functionality.
+
+Copyright (C) 2024 Canonical Ltd.
+
+Authors
+  Pedro Avalos Jimenez <pedro.avalosjimenez@canonical.com>
+
+This program is free software: you can redistribute it and/or modify
+it under the terms of the GNU General Public License version 3,
+as published by the Free Software Foundation.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+"""
+
+import argparse
+import logging
+import os
+import shlex
+import subprocess
+import time
+import urllib
+import uuid
+from typing import Dict, List, Optional
+from urllib.parse import urlparse
+
+try:
+    import distro
+
+    if distro.id() == "ubuntu-core":
+        RELEASE = "{}.04".format(distro.version())
+        CODENAME = "focal"
+        if distro.version() == "18":
+            CODENAME = "bionic"
+        elif distro.version() == "16":
+            CODENAME = "xenial"
+    else:
+        RELEASE = distro.version()
+        CODENAME = distro.codename().split()[0].lower()
+except ImportError:
+    import lsb_release  # type: ignore
+
+    RELEASE = lsb_release.get_distro_information()["RELEASE"]
+    CODENAME = lsb_release.get_distro_information()["CODENAME"]
+
+ARCH = os.uname().machine
+
+
+# XXX: The repository part of this should go away if we package mixbench
+NVIDIA_URL = "https://developer.download.nvidia.com/compute/cuda/repos/ubuntu{}/{}".format(  # noqa: E501
+    RELEASE.replace(".", ""), ARCH
+)
+GPU_VENDORS = {
+    "nvidia": {
+        "repos": [
+            {
+                "name": "cuda",
+                "repo_line": "{} /".format(NVIDIA_URL),
+                "gpg_url": "{}/3bf863cc.pub".format(NVIDIA_URL),
+                "gpg_fingerprint": "EB693B3035CD5710E231E123A4B469963BF863CC",
+                "pinfile": "{}/cuda-ubuntu{}.pin".format(
+                    NVIDIA_URL, RELEASE.replace(".", "")
+                ),
+            }
+        ],
+    },
+}
+GPU_RUNS = 20
+GPU_THRESHOLD_SEC = 12.0  # TODO: finetune
```

Comment:
Please document that these can be overridden via an envvar or a parameter to the script